File: src/main/scala/kim/dencka/scala/easy/collection/SingleNumber.scala
package kim.dencka.scala.easy.collection

object SingleNumber extends App {

  val arr = Array(2, 2, 1)
  val arr2 = Array(4, 1, 2, 1, 2)

  // FIXME MUST RESOLVE (SOLUTION ACCEPTED BUT I THINK IT TAKES TOO MUCH TIME AND MEMORY)

  def singleNumber(nums: Array[Int]): Int = {
    @annotation.tailrec
    def run(arr: Array[Int]): Int = arr match {
      case ar if nums.count(_ == ar.head) == 1 => ar.head
      case ar => run(ar.tail)
    }
    run(nums.distinct)
  }

  println(singleNumber(arr))
  println(singleNumber(arr2))

}

object MustResolve

File: src/main/scala/kim/dencka/java/easy/collection/SingleNumber.java
package kim.dencka.java.easy.collection;

public class SingleNumber {
    public static void main(String[] args) {
        int[] nums = {4, 1, 2, 1, 2};
        int[] nums2 = {2, 2, 1};
        int[] nums3 = {-3, -3, -2, -2, -1, 0, 0, 1, 1, 2, 2, 3, 3};
        System.out.println(singleNumber(nums));
        System.out.println(singleNumber(nums2));
        System.out.println(singleNumber(nums3));
    }

    public static int singleNumber(int[] nums) {
        int xor = nums[0];
        for (int i = 1; i < nums.length; i++) xor ^= nums[i];
        return xor;
    }
}
